### PR TITLE
modularity: Document the ability to use non-files in imports

### PR DESCRIPTION
**Pull request by Baughn**
###### Motivation for this change
There are no examples of using anything but files in the imports list. Users might be confused into thinking that's all that's possible.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: nixos/doc/manual/configuration/modularity.xml [hunk 1]
```diff
@@ -127,4 +127,23 @@ nix-repl> map (x: x.hostName) config.<xref linkend="opt-services.httpd.virtualHo
 [ "example.org" "example.gov" ]
 </screen>
  </para>
+
+ <para>
+   While abstracting your configuration, you may find it useful to generate
+   modules using code, instead of writing files. This is also an option; the
```
Comment (markuskowa) [inline nixos/doc/manual/configuration/modularity.xml]:
```suggestion
   modules using code, instead of writing files. The example
```

**Review by markuskowa — COMMENTED** [review]
This is indeed a useful hint (I did not realize that this is an option).

**Comment by markuskowa** [comment]
Can you please squash the commits.

**Comment by Baughn** [comment]
In a few hours. ...can't you use the "squash and merge" option?

**Comment by markuskowa** [comment]
It seems like it was designed to also accept configuration sets and not just paths. That should be documented and not just be a hidden feature? I can see your point that it makes things harder to debug but should that not be the users choice in the end? 
I personally find that feature quite useful for defining nixops deployments, where machines share some configurations but not so much that I would want to put in a separate file.

**Comment by domenkozar** [comment]
It's a slippery slope, you'll soon regret saving those 5s once you're deep debugging some coercion gone wrong, without having the line number where it happened.

I think we should at very least make this clear (in red) in the documentation.

**Comment by markuskowa** [comment]
I agree it should be mentioned in the documentation.
However, from a technical point it is not clear to me why or when nix loses track of the line numbers. I ran into this type of problem many times without using the above mentioned construct. Is that mentioned somewhere in the documentation?

**Comment by domenkozar** [comment]
You only run into this if you use nixos modules imports without filenames.

**Comment by domenkozar** [comment]
Maybe @edolstra can explain precisely, I don't have time to really research this, I'm just warning this is going to create more pain than gain as it currently stands if documented.

**Comment by edolstra** [comment]
Well, this feature is used in lots of places, so documenting it makes sense, even if it's not really a good idea to use.

**Comment by Baughn** [comment]
For instance, I'm using it to generate NixOS systems for a number of machines, most of which differ only in their hostname.

Is this a bad idea? Should I do it some other way? I don't want to generate configuration.nix files for each of them; I'd rather have the machines all defined in a single file, one line per machine.

**Pull request merged**